polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -28,305 +31,365 @@ struct ParallelLICM : public ParallelLICMBase<ParallelLICM> {
+    return writeResources;
+  };
+  const SmallVector<SideEffects::Resource *> &getFreeResources() const {
+    return freeResources;
+  };
+  const SmallVector<SideEffects::Resource *> &getAllocateResources() const {
+    return allocateResources;
+  };
+
+private:
+  SmallVector<SideEffects::Resource *> readResources;
+  SmallVector<SideEffects::Resource *> writeResources;
+  SmallVector<SideEffects::Resource *> freeResources;
+  SmallVector<SideEffects::Resource *> allocateResources;
+};
+
+/// Returns true if the Operation \p op can be hoisted out of the given loop \p
+/// loop. The \p willBeMoved argument represents operations that are known to be
+/// loop invariant (and therefore will be moved outside of the loop).
+template <typename T, typename = std::enable_if_t<llvm::is_one_of<
+                          T, scf::ParallelOp, AffineParallelOp>::value>>
+static bool canBeHoisted(Operation &op, T loop,
+                         const SmallPtrSetImpl<Operation *> &willBeMoved) {
+  LLVM_DEBUG(llvm::dbgs() << "Analyzing: " << op << "\n");
+
+  // Returns true if the given value can be moved outside of the loop, and false
+  // otherwise. A value can be moved outside of the loop if its operand are
+  // defined by operations that can themselves be moved, or are already outside
+  // of the loop.
+  auto canBeMoved = [&](Value value) {
+    if (auto BA = value.dyn_cast<BlockArgument>())
+      if (willBeMoved.count(BA.getOwner()->getParentOp()))
+        return true;
+    Operation *definingOp = value.getDefiningOp();
+    if ((definingOp && willBeMoved.count(definingOp)) ||
+        loop.isDefinedOutsideOfLoop(value))
+      return true;
+    return false;
+  };
+
+  // Ensure operands can be hoisted.
+  if (llvm::any_of(op.getOperands(),
+                   [&](Value value) { return !canBeMoved(value); })) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "cannot be hoisted: operand(s) can't be hoisted\n");
+    return false;
+  }
+
+  // If the operation has not side effects it can be hoisted.
+  if (isMemoryEffectFree(&op)) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "can be hoisted: operation has no side effects\n");
+    return true;
+  }
+
+  // Operations with unknown side effects cannot be hoisted.
+  if (!isa<MemoryEffectOpInterface>(op) &&
+      !op.hasTrait<OpTrait::HasRecursiveMemoryEffects>()) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "cannot be hoisted: unknown side effects\n");
+    return false;
+  }
+
+  // Do not hoist operations that allocate memory.
+  const OperationMemoryEffects memoryEffects(op);
+  if (memoryEffects.allocatesMemory()) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "cannot be hoisted: operation allocates memory\n");
+    return false;
+  }
+
+  // Determine whether the given operation has memory effects that 'conflict'
+  // with other operations in the same loop.
+  std::function<bool(Operation &, LoopLikeOpInterface)> hasConflictsInLoop =
```

Comment:
```suggestion
  const auto hasConflictsInLoop =
```
Closures are lower-weight than `std::function`

Reply:
My bad, didn't see it